msal/src/main/java/com/microsoft/identity/client/NativeAuthPublicClientApplication.kt

```diff
@@ -111,5 +185,381 @@ class NativeAuthPublicClientApplication(
+                callback.onError(e)
+            }
+        }
+    }
+
+    /**
+     * Retrieve the current signed in account from cache; Kotlin coroutines variant.
+     *
+     * @return [com.microsoft.identity.client.statemachine.states.AccountResult] if there is a signed in account, null otherwise.
+     */
+    override suspend fun getCurrentAccount(): AccountResult? {
+        return withContext(Dispatchers.IO) {
+            val account = getCurrentAccountInternal(nativeAuthConfig)
+            return@withContext if (account != null) {
+                AccountResult.createFromAccountResult(
+                    account = account,
+                    config = nativeAuthConfig
+                )
+            } else {
+                null
+            }
+        }
+    }
+
+    interface SignInCallback : Callback<SignInResult>
+
+    /**
+     * Sign in a user with a given username; callback variant.
+     *
+     * @param username username of the account to sign in.
+     * @param scopes (Optional) scopes to request during the sign in.
+     * @param callback [com.microsoft.identity.client.NativeAuthPublicClientApplication.SignInCallback] to receive the result.
+     * @return [com.microsoft.identity.client.statemachine.results.SignInResult] see detailed possible return state under the object.
+     * @throws [MsalException] if an account is already signed in.
+     */
+    override fun signIn(username: String, scopes: List<String>?, callback: SignInCallback) {
+        LogSession.logMethodCall(TAG, "${TAG}.signIn(username: String, scopes: List<String>?, callback: SignInCallback)")
+        pcaScope.launch {
+            try {
+                val result = signIn(username, scopes)
+                callback.onResult(result)
+            } catch (e: MsalException) {
+                Logger.error(TAG, "Exception thrown in signIn", e)
+                callback.onError(e)
+            }
+        }
+    }
+
+    /**
+     * Sign in a user with a given username; Kotlin coroutines variant.
+     *
+     * @param username username of the account to sign in.
+     * @param scopes (Optional) scopes to request during the sign in.
+     * @return [com.microsoft.identity.client.statemachine.results.SignInResult] see detailed possible return state under the object.
+     * @throws [MsalException] if an account is already signed in.
+     */
+    override suspend fun signIn(
+        username: String,
+        scopes: List<String>?
+    ): SignInResult {
+        return withContext(Dispatchers.IO) {
+            LogSession.logMethodCall(TAG, "${TAG}.signIn")
+
+            val doesAccountExist = checkForPersistedAccount().get()
+            if (doesAccountExist) {
+                Logger.error(
+                    TAG,
+                    "An account is already signed in.",
+                    null
+                )
+                throw MsalClientException(
+                    MsalClientException.INVALID_PARAMETER,
```

Comment:
Is INVALID_PARAMETER the right error in this case?

Reply:
This is the most suitable error from existing codes.